tidb/src/main/java/io/tidb/bigdata/tidb/RowBuffer.java

```diff
@@ -117,32 +123,47 @@ public boolean add(Row row) {
         return true;
       }
       List<List<Object>> values = new ArrayList<>();
-      for (Pair<List<Integer>, Set<List<Object>>> pair : uniqueIndexValues) {
+      boolean result = true;
+      for (Pair<List<Integer>, Map<List<Object>, Row>> pair : uniqueIndexValues) {
         List<Integer> indexColumns = pair.first;
-        Set<List<Object>> indexValues = pair.second;
+        Map<List<Object>, Row> indexValues = pair.second;
         List<Object> indexValue =
-            indexColumns.stream().map(i -> row.get(i, null)).collect(Collectors.toList());
-        if (indexValues.contains(indexValue)) {
-          return false;
+            Collections.unmodifiableList(
+                indexColumns.stream().map(i -> row.get(i, null)).collect(Collectors.toList()));
+        if (indexValues.containsKey(indexValue)) {
+          result = false;
```

Comment:
What is the result actually mean?

Reply:
the result means whether the row has a conflict with the rows in rowbuffer.
if duplication is false and result is false, exception will be thrown